flow/Histogram.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Histogram.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FLOW_HISTOGRAM_H
+#define FLOW_HISTOGRAM_H
+#pragma once
+
+#include <string>
+#include <map>
+
+#ifdef _WIN32
+#include <intrin.h>
+#pragma intrinsic(_BitScanReverse)
+#endif
+
+class Histogram;
+
+class HistogramRegistry {
+    public:
+        void registerHistogram(Histogram * h);
+        void unregisterHistogram(Histogram * h);
+        void logReport();
+    private:
+        std::map<std::string, Histogram*> histograms;
+};
+
+// TODO: This should be scoped properly for simulation (instead of just having all the "machines" share one histogram namespace)
+HistogramRegistry & GetHistogramRegistry();
+
+class Histogram {
```

Comment:
I'd like to add some TEST_CASE's for this, but do we have a good way of checking that we're logging expected values?

As implemented, this isn't a great way to make the inferences you're asking for, since the histogram's memory gets wiped each log period.  That'd cause the computed p90 to be extremely noisy.

Reply:
Though, we could have it report percentiles in the report, which would be easier to present in things like status json.  The trick would be to present the previous histogram's percentiles.  That way, the measurements wouldn't be noisy right after we wipe histogram state.  However, this would give a trend line with a lot of quantization error; it would always be rounded to a power of two.